pyccel/parser/syntax/openmp.py

```diff
@@ -31,160 +35,686 @@ def __init__(self, **kwargs):
         """
         self.stmt = kwargs.pop('stmt')
 
-        super(OpenmpStmt, self).__init__(**kwargs)
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
             print("> OpenmpStmt: expr")
 
         stmt = self.stmt
-        if isinstance(stmt, OmpEndClause):
-            return stmt.expr
-        elif isinstance(stmt, OmpParallelConstruct):
-            return stmt.expr
-        elif isinstance(stmt, OmpLoopConstruct):
-            return stmt.expr
-        elif isinstance(stmt, OmpSingleConstruct):
+        if isinstance(stmt, (OmpEndClause, OmpParallelConstruct, OmpLoopConstruct,
+                             OmpSingleConstruct, OmpCriticalConstruct, OmpBarrierConstruct,
+                             OmpMasterConstruct, OmpMaskedConstruct, OmpTaskLoopConstruct,
+                             OmpSimdConstruct, OmpAtomicConstruct, OmpTaskWaitConstruct,
+                             OmpTaskyieldConstruct, OmpTaskConstruct, OmpFlushConstruct,
+                             OmpCancelConstruct, OmpTargetConstruct, OmpTeamsConstruct,
+                             OmpDistributeConstruct, OmpSectionConstruct, OmpSectionsConstruct)):
             return stmt.expr
         else:
             raise TypeError('Wrong stmt for OpenmpStmt')
 
+def check_get_clauses(name, valid_clauses, clauses):
+    """
+    Function check if the clauses are correct for a given Construct
+    """
+    txt = ''
+    for clause in clauses:
+        if isinstance(clause, valid_clauses):
+            txt = '{0} {1}'.format(txt, clause.expr)
+        else:
+            msg = "Wrong clause " + type(clause).__name__
+            raise TypeError(msg)
+    return txt
+
 class OmpParallelConstruct(BasicStmt):
-    """Class representing a ."""
+    """Class representing a parallel construct."""
+    def __init__(self, **kwargs):
+        """
+        """
+        self.clauses  = kwargs.pop('clauses')
+        self.combined = kwargs.pop('combined')
+
+        _valid_clauses = _valid_parallel_clauses
+
+        com = None
+        if isinstance(self.combined, OmpForSimd):
+            _valid_clauses = _valid_clauses + _valid_loop_clauses
+            if 'simd' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_simd_clauses
+            com = self.combined.expr
+        if isinstance(self.combined, OmpMaskedTaskloop):
+            _valid_clauses = _valid_clauses + (OmpFilter,)
+            if 'simd' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_simd_clauses
```

Comment:
Here you do `_valid_clauses = _valid_clauses + _valid_simd_clauses` if `'simd' in self.combined.expr` and `self.combined` is one of: `OmpForSimd`, `OmpMaskedTaskLoop`. Can other classes have `'simd' in self.combined.expr`?

Reply:
yes, but not all, only construct that can be combined and can have `simd` as a combined construct, so not all other classes can have it in `self.combined.expr`

Reply:
In that case why do you need all the `isinstance` checks?
Why not just do:
```
_valid_clauses = self._valid_clauses
if 'simd' in self.combined.expr:
    _valid_clauses += _valid_simd_clauses
if 'taskloop' in self.combined.expr:
    _valid_clauses += _valid_taskloop_clauses
```
?

Reply:
but `self.combined` can be `None`, then if I check for `if 'simd' in self.combined.expr:` I will get a python error object has no attribute 'expr'

Reply:
Option 1:
Create a class `OmpParallelCombinedConstruct` for objects which have a `combined` property
Option 2:
```
_valid_clauses = self._valid_clauses
if 'simd' in self.combined:
    if 'simd' in self.combined.expr:
        _valid_clauses += _valid_simd_clauses
    if 'taskloop' in self.combined.expr:
        _valid_clauses += _valid_taskloop_clauses
```

Reply:
im not sure i understand the first option as i know in `textx` every class represent a rule, if i work with option 2 i need to add more if to check `if 'for' in self.combined.expr` because when i do `if isinstance(self.combined, OmpForSimd):` I'm sure here that `self.combined.expr` contains `for` and the same for the others

Reply:
Surely in `textx` every _imported_ class represents a rule? But the imported classes can still inherit from other classes that `textx` doesn't need to know about correctly?

Reply:
As I understand it (bearing in mind that I have never worked with `textx`), `textx` constructs a metamodel from the file `pyccel/parser/grammar/openmp.tx`. This file tells `textx` which objects need to be created depending on which words are encountered. It doesn't tell `textx` anything about the classes that these objects might inherit from

Reply:
>if i work with option 2 i need to add more if to check if 'for' in self.combined.expr because when i do if isinstance(self.combined, OmpForSimd): I'm sure here that self.combined.expr contains for and the same for the others

Hmm, maybe if combined objects also had `_valid_clauses`?